internal/transport/transport_test.go

```diff
@@ -2746,3 +2746,80 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
 		t.Errorf("Context timed out")
 	}
 }
+
+// Test that in the event of a graceful client transport shutdown
+// ,  i.e., clientTransport.Close(), if the GOAWAY write is not
+// finished within specified time due to network hang, client
+// should still close without waiting for too long.
+func (s) TestClientCloseReturnsEarlyWhenGoAwayWriteHangs(t *testing.T) {
+	// Override timer for writing GOAWAY to 0 so that the connection write
+	// always times out. It is equivalent of real network hang when conn
+	// write for goaway doesn't finish in specified deadline
+	origGoAwayLoopyTimeout := goAwayLoopyWriterTimeout
+	goAwayLoopyWriterTimeout = 0
+	defer func() {
+		goAwayLoopyWriterTimeout = origGoAwayLoopyTimeout
+	}()
+
+	// Create a server.
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening: %v", err)
+	}
+	defer lis.Close()
+	// errorCh verifies that desired GOAWAY not received by server
+	errorCh := make(chan error, 1)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	// Launch the server and handle HTTP/2 connections, specifically
+	// focusing on the reception and interpretation of GOAWAY frames
+	// sent from client.
+	go func() {
+		defer close(errorCh)
+		conn, err := lis.Accept()
+		if err != nil {
+			t.Errorf("Error while accepting: %v", err)
+		}
+		defer conn.Close()
+		framer := http2.NewFramer(conn, conn)
+		if err := framer.WriteSettings(); err != nil {
+			t.Errorf("Error while writing settings %v", err)
+			return
+		}
+
+		frame, err := framer.ReadFrame()
+		if err != nil {
+			return
+		}
+		switch fr := frame.(type) {
+		case *http2.GoAwayFrame:
+			if fr.ErrCode == http2.ErrCodeNo {
+				t.Logf("Received goAway frame from client")
+				return
+			}
+			errorCh <- fmt.Errorf("received unexpected goAway frame: %v", err)
+		default:
+			errorCh <- fmt.Errorf("server received a frame other than GOAWAY: %v", err)
+			return
+		}
+	}()
+
+	ct, err := NewClientTransport(ctx, ctx, resolver.Address{Addr: lis.Addr().String()}, ConnectOptions{}, func(GoAwayReason) {})
+	if err != nil {
+		t.Fatalf("Error while creating client transport: %v", err)
+	}
+	_, err = ct.NewStream(ctx, &CallHdr{})
+	if err != nil {
+		t.Fatalf("failed to open stream: %v", err)
+	}
+	// ct.Close will try to send the GOAWAY to server but conn.
+	// Write will time out due to goAwayLoopyWriterTimeout being
+	// 0. It is equivalent of a network hang when GOAWAY doesn't
+	// finish within specified deadline.
+	ct.Close(errors.New("manually closed by client"))
```

Comment:
If the call to `ct.Close` hangs (i.e the bug that we are trying to fix manifests itself), then we will get get to the `select` below.